app/controllers/multifactor_auths_controller.rb

```diff
@@ -37,6 +43,50 @@ def update
     end
   end
 
+  def update
+    unless valid_mfa_level?
+      redirect_to edit_settings_path, flash: { error: t(".invalid_level") }
+      return
+    end
+
+    session[:level] = level_param
+    @user = current_user
+
+    setup_mfa_authentication
+    setup_webauthn_authentication
+
+    create_new_mfa_expiry
+
+    render template: "multifactor_auths/mfa_prompt"
+  end
+
+  def mfa_update
+    unless current_user.totp_enabled?
+      redirect_to edit_settings_path, flash: { error: t("multifactor_auths.require_totp_enabled") }
+      return
+    end
+
+    if current_user.ui_mfa_verified?(params[:otp])
+      update_level_and_redirect
+    else
+      redirect_to edit_settings_path, flash: { error: t("multifactor_auths.incorrect_otp") }
+    end
+  end
+
+  def webauthn_update
+    @user = current_user
+    unless @user.webauthn_enabled?
+      redirect_to edit_settings_path, flash: { error: t("multifactor_auths.require_webauthn_enabled") }
+      return
+    end
+
+    if webauthn_credential_verified?
```

Comment:
I was able to condense this quite a bit since `WebauthnVerifiable` extract most of the verifying credential logic.